### PR TITLE
multiple device refresh improvements

### PR DESCRIPTION
**Pull request by husobee**
### Summary

updates to delete credentials to only hard delete where request id matches item id

### Type of Change

- [ ] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: services/skus/credentials.go [hunk 1]
```diff
@@ -720,7 +720,20 @@ func (s *Service) DeleteOrderCreds(ctx context.Context, orderID uuid.UUID, isSig
 	}
 
 	if doTlv2 {
-		if err := s.Datastore.DeleteTimeLimitedV2OrderCredsByOrderTx(ctx, tx, orderID); err != nil {
+		var itemIDs []*uuid.UUID
+		if reqID == nil {
+			// legacy, so put all of the orders' items in a list to be hard deleted
+			for _, v := range order.Items {
+				itemIDs = append(itemIDs, &v.ID)
```
Comment (clD11) [inline services/skus/credentials.go:727]: Will this not result in the same value for every item given its taking a pointer in the loop?

File: services/skus/controllers.go [hunk 2]
```diff
@@ -693,7 +735,7 @@ func DeleteOrderCreds(service *Service) handlers.AppHandler {
 		}
 
 		isSigned := r.URL.Query().Get("isSigned") == "true"
-		if err := service.DeleteOrderCreds(ctx, id, isSigned); err != nil {
+		if err := service.DeleteOrderCreds(ctx, id, nil, isSigned); err != nil {
```
Comment (clD11) [inline services/skus/controllers.go:738]: Can we avoid passing nil into methods or functions it leads to nil checks and harder to maintain code

File: services/skus/order.go [hunk 3]
```diff
@@ -198,5 +198,5 @@ func (s *Service) RenewOrder(ctx context.Context, orderID uuid.UUID) error {
 		return fmt.Errorf("failed to set order status to paid: %w", err)
 	}
 
-	return s.DeleteOrderCreds(ctx, orderID, true)
+	return s.DeleteOrderCreds(ctx, orderID, nil, true)
```
Comment (clD11) [inline services/skus/order.go:201]: Same comment as about passing nil

File: services/skus/credentials.go [hunk 4]
```diff
@@ -688,7 +688,7 @@ func (s *SigningOrderResultErrorHandler) Handle(ctx context.Context, message kaf
 // - create repos for credentials;
 // - move the corresponding methods there;
 // - make those methods work on per-item basis.
-func (s *Service) DeleteOrderCreds(ctx context.Context, orderID uuid.UUID, isSigned bool) error {
+func (s *Service) DeleteOrderCreds(ctx context.Context, orderID uuid.UUID, reqID *uuid.UUID, isSigned bool) error {
```
Comment (clD11) [inline services/skus/credentials.go:691]: question: is there a specific reason we are passing `reqID` as a pointer should this not be a value?

It suggest that it is going to be modified in some way

File: services/skus/controllers.go [hunk 5]
```diff
@@ -588,6 +589,47 @@ func CreateOrderCreds(svc *Service) handlers.AppHandler {
 	}
 }
 
+// deleteItemCreds handles requests for deleting credentials for an item.
+func deleteItemCreds(svc *Service) handlers.AppHandler {
+	return func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		ctx := r.Context()
+		lg := logging.Logger(ctx, "skus.deleteItemCreds")
```
Comment (clD11) [inline services/skus/controllers.go:596]: suggestion: inline with the new approach pass the module to as the second argument and define the func separately. e.g.

`logging.Logger(ctx, "skus").Str("func", "deleteItemCreds").Logger()`

File: services/skus/controllers.go [hunk 6]
```diff
@@ -588,6 +589,47 @@ func CreateOrderCreds(svc *Service) handlers.AppHandler {
...
+
+		orderID := &inputs.ID{}
+		if err := inputs.DecodeAndValidateString(ctx, orderID, chi.URLParamFromCtx(ctx, "orderID")); err != nil {
```
Comment (clD11) [inline services/skus/controllers.go:599]: question: are we not trying to move away from this approach of using the `DecodeAndValidateString` given it adds an extra layer of abstraction which seems unnecessary?

File: services/skus/controllers.go [hunk 7]
```diff
@@ -588,6 +589,47 @@ func CreateOrderCreds(svc *Service) handlers.AppHandler {
...
+			return handlers.ValidationError("Error validating request url parameter", map[string]interface{}{
+				"requestID": err.Error(),
+			})
+		}
+
+		isSigned := r.URL.Query().Get("isSigned") == "true"
+
+		if err := svc.DeleteOrderCreds(ctx, *orderID.UUID(), reqID.UUID(), isSigned); err != nil {
+			lg.Error().Err(err).Msg("failed to delete the order credentials")
+			return handlers.WrapError(err, "Error deleting credentials", http.StatusBadRequest)
```
Comment (clD11) [inline services/skus/controllers.go:626]: question: is there a reason we are returning a bad request opposed to an internal server error or a status not found if we cant find the order. It does not seem 400 is the right error at this point in the processing.

**Comment by clD11** [comment]
Could we add some tests to cover the new functionality/endpoints?

File: services/skus/controllers.go [hunk 8]
```diff
@@ -588,6 +589,47 @@ func CreateOrderCreds(svc *Service) handlers.AppHandler {
...
+			lg.Error().Err(err).Msg("failed to validate order id")
+			return handlers.ValidationError("Error validating request url parameter", map[string]interface{}{
+				"orderID": err.Error(),
+			})
+		}
+
+		itemID := &inputs.ID{}
+		if err := inputs.DecodeAndValidateString(ctx, itemID, chi.URLParamFromCtx(ctx, "itemID")); err != nil {
+			lg.Error().Err(err).Msg("failed to validate item id")
+			return handlers.ValidationError("Error validating request url parameter", map[string]interface{}{
```
Comment (clD11) [inline services/skus/controllers.go:609]: suggestion: improve the readability of the error message.

If we look the [`handlers.ValidationError()`](https://github.com/brave-intl/bat-go/blob/9f0b7aa487086dd4b0769f437a2ae5114a29a520/libs/handlers/handlers.go#L116) function it concats the passed in message with `Error validating` so we end up with a message like `Error validating Error validating request url parameter` which does not read very well. 

Can this be changed?





`Error validating Error validating request url parameter`

File: services/skus/controllers.go (hunk 2 above)
Comment (pavelbrm) [reply services/skus/controllers.go:738]: Plus a lot to this. This does not look good to me.

Instead, declare that as `uuid.UUID`, and pass `uuid.Nil` to indicate the emptiness of the request id.

File: services/skus/credentials.go (hunk 4 above)
Comment (pavelbrm) [reply services/skus/credentials.go:691]: Plus a lot to this.

Suggestion: make it `uuid.UUID` and make the caller pass either a normal uuid, or `uuid.Nil`.

File: services/skus/credentials.go (hunk 1 above)
Comment (pavelbrm) [reply services/skus/credentials.go:727]: Also, it's unclear why use a slice of pointers to uuids. Under the hood, a UUID is a small array (not even a slice), which is OK to pass around by value, and `uuid.Nil` serves as a way to pass/indicate emptiness.

Additionally, `itemIDs` can be preallocated as we know the number of items upfront.

As for the `reqID`, the check should become `if uuid.Equal(reqID, uuid.Nil)`.

File: services/skus/datastore.go [hunk 9]
```diff
@@ -83,7 +83,7 @@ type Datastore interface {
 	AreTimeLimitedV2CredsSubmitted(ctx context.Context, blindedCreds ...string) (bool, error)
 	GetTimeLimitedV2OrderCredsByOrder(orderID uuid.UUID) (*TimeLimitedV2Creds, error)
 	GetTLV2Creds(ctx context.Context, dbi sqlx.QueryerContext, ordID, itemID, reqID uuid.UUID) (*TimeLimitedV2Creds, error)
-	DeleteTimeLimitedV2OrderCredsByOrderTx(ctx context.Context, tx *sqlx.Tx, orderID uuid.UUID) error
+	DeleteTimeLimitedV2OrderCredsByOrderTx(ctx context.Context, tx *sqlx.Tx, orderID uuid.UUID, itemIDs ...*uuid.UUID) error
```
Comment (pavelbrm) [inline services/skus/datastore.go:86]: Issue: values, not pointers for itemIDs, please.

File: services/skus/datastore.go [hunk 10]
```diff
@@ -796,8 +796,16 @@ func (pg *Postgres) DeleteSingleUseOrderCredsByOrderTx(ctx context.Context, tx *
 
 // DeleteTimeLimitedV2OrderCredsByOrderTx performs a hard delete for all time limited v2 order
 // credentials for a given OrderID.
-func (pg *Postgres) DeleteTimeLimitedV2OrderCredsByOrderTx(ctx context.Context, tx *sqlx.Tx, orderID uuid.UUID) error {
-	_, err := tx.ExecContext(ctx, `delete from time_limited_v2_order_creds where order_id = $1`, orderID)
+func (pg *Postgres) DeleteTimeLimitedV2OrderCredsByOrderTx(ctx context.Context, tx *sqlx.Tx, orderID uuid.UUID, itemIDs ...*uuid.UUID) error {
+	// legacy, if the request id matches the item id on the time limited v2 creds, we actually want to delete the record,
+	// otherwise we will keep these credentials here for multiple device refresh capabilities
+	query, args, err := sqlx.In("delete from time_limited_v2_order_creds where order_id = ? and item_id in (?)", orderID, itemIDs)
```
Comment (pavelbrm) [inline services/skus/datastore.go:802]: issue: question mark is a placeholder in MySQL. We use Postgres where numerated placeholders are used.

File: services/skus/order.go (hunk 3 above)
Comment (pavelbrm) [reply services/skus/order.go:201]: Plus one to this.

**Review by pavelbrm — CHANGES_REQUESTED** [review]
There are a few important bugs in the new code that need to be fixed before this is ready. At the very least, in the order of importance:
- The captured loop variable will result in the same value stored `len(slice)` times. Plus pointers.
- Pointers to UUIDs need to be replaced with values.

File: services/skus/datastore.go (hunk 10 above)
Comment (pavelbrm) [reply services/skus/datastore.go:802]: Checked the docs. Found why this should use `?`. Resolved.

**Review by pavelbrm — COMMENTED** [review]
I went ahead and expressed my feedback in #2314. Please read the description and the PR. It addresses the following aspects:
- Expresses the deletion logic so it's testable;
- Added test for the deletion logic so that we can see what's being passed down to the datastore level;
- Fixed types for UUID.

File: services/skus/credentials.go (hunk 4 above)
Comment (pavelbrm) [reply services/skus/credentials.go]: Updated by the patch #2314.

File: services/skus/order.go (hunk 3 above)
Comment (pavelbrm) [reply services/skus/order.go]: Updated by the patch #2314.

File: services/skus/controllers.go (hunk 2 above)
Comment (pavelbrm) [reply services/skus/controllers.go]: Updated by the patch #2314.

File: services/skus/credentials.go (hunk 1 above)
Comment (pavelbrm) [reply services/skus/credentials.go]: Updated by the patch #2314.

File: services/skus/datastore.go [hunk 11]
```diff
@@ -1026,6 +1036,21 @@ func (pg *Postgres) GetTLV2Creds(ctx context.Context, dbi sqlx.QueryerContext, o
 	return result, nil
 }
 
+// GetCountActiveOrderCreds returns the count of order creds currently active on an order
+func (pg *Postgres) GetCountActiveOrderCreds(ctx context.Context, dbi sqlx.ExtContext, orderID uuid.UUID) (int, error) {
+	query := `
+		select count(1) from time_limited_v2_order_creds
+		where order_id = $1 and now() > valid_from and valid_to > now() group by request_id
```
Comment (clD11) [inline services/skus/datastore.go:1043]: issue: we are using two different times.

Pass time.Now into the function.

File: services/skus/datastore_test.go [hunk 12]
```diff
@@ -151,6 +151,24 @@ func (suite *PostgresTestSuite) TestGetOrderByExternalID() {
 	}
 }
 
+func (suite *PostgresTestSuite) TestCountActiveOrderCreds_Success() {
```
Comment (clD11) [inline services/skus/datastore_test.go:154]: suggestions(non-blocking): this test doesn't follow the new pattern and we are also missing several conditions such as expired and tokens.

Add the extra conditions and refactor to the new pattern

File: services/skus/datastore_test.go [hunk 13]
```diff
@@ -151,6 +151,24 @@ func (suite *PostgresTestSuite) TestGetOrderByExternalID() {
...
+		devBraveSearchPremiumYearTimeLimited})
+
+	orderCredentials := suite.createTimeLimitedV2OrderCreds(suite.T(), ctx, devBraveFirewallVPNPremiumTimeLimited,
+		devBraveSearchPremiumYearTimeLimited)
+
+	// both order items have same orderID so can use the first element to retrieve all order creds
+	count, err := suite.storage.GetCountActiveOrderCreds(ctx, suite.storage.RawDB(), orderCredentials[0].OrderID)
+	suite.Require().NoError(err)
+
+	suite.Assert().Equal(count, 2)
```
Comment (clD11) [inline services/skus/datastore_test.go:169]: suggestion: the equal parameters are the wrong way round, the signature of equals is expected then actual.

Change this to `suite.Assert().Equal(2, count)`

File: services/skus/datastore_test.go [hunk 14]
```diff
@@ -151,6 +151,24 @@ func (suite *PostgresTestSuite) TestGetOrderByExternalID() {
...
+
+	// create paid order with two order items
+	ctx = context.WithValue(ctx, appctx.WhitelistSKUsCTXKey, []string{devBraveFirewallVPNPremiumTimeLimited,
+		devBraveSearchPremiumYearTimeLimited})
+
+	orderCredentials := suite.createTimeLimitedV2OrderCreds(suite.T(), ctx, devBraveFirewallVPNPremiumTimeLimited,
+		devBraveSearchPremiumYearTimeLimited)
+
+	// both order items have same orderID so can use the first element to retrieve all order creds
+	count, err := suite.storage.GetCountActiveOrderCreds(ctx, suite.storage.RawDB(), orderCredentials[0].OrderID)
```
Comment (clD11) [inline services/skus/datastore_test.go:166]: nit: `count` to be changed to actual to make the assertion clearer.

**Review by pavelbrm — COMMENTED** [review]
Waiting on the open comments to be addressed before doing the final review.

File: services/skus/controllers.go (hunk 5 above)
Comment (husobee) [reply services/skus/controllers.go]: addressed: https://github.com/brave-intl/bat-go/pull/2281/commits/a40fb6cdf80029cc56e66e28a0ede7d70f191587

File: services/skus/controllers.go (hunk 6 above)
Comment (husobee) [reply services/skus/controllers.go]: addressed: https://github.com/brave-intl/bat-go/pull/2281/commits/a40fb6cdf80029cc56e66e28a0ede7d70f191587

File: services/skus/controllers.go (hunk 7 above)
Comment (husobee) [reply services/skus/controllers.go]: addressed: https://github.com/brave-intl/bat-go/pull/2281/commits/a40fb6cdf80029cc56e66e28a0ede7d70f191587

File: services/skus/controllers.go (hunk 8 above)
Comment (husobee) [reply services/skus/controllers.go]: addressed: https://github.com/brave-intl/bat-go/pull/2281/commits/a40fb6cdf80029cc56e66e28a0ede7d70f191587

File: services/skus/datastore_test.go (hunk 14 above)
Comment (husobee) [reply services/skus/datastore_test.go]: addressed: https://github.com/brave-intl/bat-go/pull/2281/commits/5012c82cbba0e4b3cee709a81f4c4f6bac6e6f7a

File: services/skus/datastore_test.go (hunk 13 above)
Comment (husobee) [reply services/skus/datastore_test.go]: addressed: https://github.com/brave-intl/bat-go/pull/2281/commits/5012c82cbba0e4b3cee709a81f4c4f6bac6e6f7a

File: services/skus/datastore.go (hunk 11 above)
Comment (husobee) [reply services/skus/datastore.go]: addressed: https://github.com/brave-intl/bat-go/pull/2281/commits/ff503ecb24c7e7cdd23a13ef491fd7ae647276ed

**Review by clD11 — APPROVED** [review]

File: services/skus/controllers.go [hunk 15]
```diff
@@ -600,6 +601,42 @@ func CreateOrderCreds(svc *Service) handlers.AppHandler {
...
+
+		isSigned := r.URL.Query().Get("isSigned") == "true"
+
+		if err := svc.DeleteOrderCreds(ctx, orderID, reqID, isSigned); err != nil {
+			if errors.Is(err, model.ErrOrderNotFound) || errors.Is(err, ErrOrderHasNoItems) {
+				return handlers.WrapError(err, "order or item not found", http.StatusNotFound)
+			}
+			lg.Error().Err(err).Msg("failed to delete the order credentials")
+			return handlers.WrapError(err, "error deleting credentials", http.StatusInternalServerError)
+		}
```
Comment (pavelbrm) [inline services/skus/controllers.go:634]: question: should the `errMaxActiveOrderCredsExceeded` error be handled somehow? Does need a non-500 response? Will the client have to be able to understand this case?

File: services/skus/datastore.go [hunk 16]
```diff
@@ -1026,6 +1036,21 @@ func (pg *Postgres) GetTLV2Creds(ctx context.Context, dbi sqlx.QueryerContext, o
 	return result, nil
 }
 
+// GetCountActiveOrderCreds returns the count of order creds currently active on an order
+func (pg *Postgres) GetCountActiveOrderCreds(ctx context.Context, dbi sqlx.ExtContext, orderID uuid.UUID, now time.Time) (int, error) {
+	query := `
```
Comment (pavelbrm) [inline services/skus/datastore.go:1041]: non-blocking nit: no reason for this to be a `var`, should be a `const`.

File: services/skus/controllers.go (hunk 15 above)
Comment (pavelbrm) [reply services/skus/controllers.go]: Addressed in https://github.com/brave-intl/bat-go/pull/2325.

File: services/skus/datastore.go (hunk 16 above)
Comment (pavelbrm) [reply services/skus/datastore.go]: Addressed in https://github.com/brave-intl/bat-go/pull/2325.

**Pull request merged**